sagan-common/src/main/java/sagan/projects/ProjectRelease.java

```diff
@@ -183,44 +187,75 @@ public ProjectRelease createWithVersionPattern() {
         return release;
     }
 
+    private static final Map<String, Integer> SPECIAL_MEANINGS =
+            ImmutableMap.of("dev", -1, "rc", 1, "release", 2, "final", 3);
+
+    /**
+     * Adapted from Gradle's StaticVersionComparator
+     */
     @Override
     public int compareTo(ProjectRelease other) {
-        if (this.getVersion() == null || other.getVersion() == null) {
-            return 0;
-        }
-
-        Pattern pattern = Pattern.compile("([0-9]+)");
+        if(other == null) return -1;
 
-        Matcher thisMatcher = pattern.matcher(this.getVersion());
-        Matcher otherMatcher = pattern.matcher(other.getVersion());
+        Version version1 = Version.build(getVersion());
+        Version version2 = Version.build(other.getVersion());
 
-        int versionDiff = 0;
-
-        boolean thisFind = thisMatcher.find();
-        boolean otherFind = otherMatcher.find();
+        if(version1 == null && version2 == null) {
+            return 0;
+        }
+        else if(version1 == null) {
+            return -1;
+        }
+        else if(version2 == null) {
+            return 1;
+        }
+        if (version1.equals(version2)) {
+            return 0;
+        }
 
-        while (thisFind && otherFind) {
-            int thisFoundVersion = Integer.parseInt(thisMatcher.group(0));
-            int otherFoundVersion = Integer.parseInt(otherMatcher.group(0));
-            versionDiff = otherFoundVersion - thisFoundVersion;
+        String[] parts1 = version1.getParts();
+        String[] parts2 = version2.getParts();
 
-            if (versionDiff != 0) {
-                return versionDiff;
+        int i = 0;
+        for (; i < parts1.length && i < parts2.length; i++) {
+            if (parts1[i].equals(parts2[i])) {
+                continue;
             }
-
-            thisFind = thisMatcher.find();
-            otherFind = otherMatcher.find();
+            boolean is1Number = isNumber(parts1[i]);
+            boolean is2Number = isNumber(parts2[i]);
+            if (is1Number && !is2Number) {
```

Comment:
Now I'm pretty far down the rabbit hole on understanding what `is1Number` is. Do you mean `isThisANumber` and `isThatANumber`?